### PR TITLE
Add unit tests for shared prefix masked attention with `torch.FlexAttention`

### PR DESCRIPTION
**Pull request by austin362667**
## Summary

> TLDR of https://github.com/linkedin/Liger-Kernel/issues/476: The shared prefix attention mask is an optimization for paired-preference alignment training.

To pave the way for https://github.com/linkedin/Liger-Kernel/issues/476, this PR aims to set up basic unit tests of flex attn with casual and shared prefix mask. 

## Benchmarks

1. Casual Attention Mask (Flash Attention 2 vs. Torch Scaled Dot Product Attention vs. FlexAttention)

image

3. Shared Prefix Attention Mask (Flash Attention 2 vs. Torch Scaled Dot Product Attention vs. FlexAttention)

image

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by austin362667** [comment]
The FlexAttention correctness check failed only on AMD GPU might be relevant to this [ISSUE](https://github.com/pytorch/pytorch/issues/138300) and [PR](https://github.com/pytorch/pytorch/pull/140172).

**Comment by austin362667** [comment]
> The FlexAttention correctness check failed only on AMD GPU might be related to this https://github.com/pytorch/pytorch/issues/138300 and https://github.com/pytorch/pytorch/pull/140172.

Maybe this is relevant: https://github.com/linkedin/Liger-Kernel/pull/506/files. It looks like PyTorch ROCm 6.2 is not as solid as PyTorch ROCm 6.3

**Review by lancerts — APPROVED** [review]

**Comment by shivam15s** [comment]
@austin362667 could you take a look at why backward compatibility tests fail for this?
